pkg/lib/encoding/encoding.go

```diff
@@ -0,0 +1,95 @@
+package encoding
+
+import (
+	"bytes"
+	"io"
+
+	"compress/gzip"
+	"encoding/base64"
+)
+
+// EncodeGzipBase64 applies gzip compression to the given bytes,
+// followed by base64 encoding.
+func EncodeGzipBase64(data []byte) ([]byte, error) {
+	gzipped, err := gzipEncode(data)
+	if err != nil {
+		return nil, err
+	}
+
+	base64d, err := base64Encode(gzipped)
+	if err != nil {
+		return nil, err
+	}
+
+	return base64d, nil
+}
+
+// DecodeGzipBase64 applies base64 decoding to the given bytes,
+// followed by gzip decompression.
+func DecodeGzipBase64(data []byte) ([]byte, error) {
+	unbase64d, err := base64Decode(data)
+	if err != nil {
+		return nil, err
+	}
+
+	ungzipped, err := gzipDecode(unbase64d)
+	if err != nil {
+		return nil, err
+	}
+
+	return ungzipped, nil
+}
+
+func gzipEncode(data []byte) ([]byte, error) {
+	buf := &bytes.Buffer{}
+
+	writer := gzip.NewWriter(buf)
+	_, err := writer.Write(data)
+	if err != nil {
+		writer.Close()
+		return nil, err
+	}
+
+	// Ensure bytes are flushed to the underlying buffer
+	writer.Close()
+	return buf.Bytes(), nil
+}
+
+func gzipDecode(data []byte) ([]byte, error) {
+	zBuf := bytes.NewBuffer(data)
+
+	reader, err := gzip.NewReader(zBuf)
+	if err != nil {
+		return nil, err
+	}
+	defer reader.Close()
+
+	buf := &bytes.Buffer{}
+	_, err = io.Copy(buf, reader)
```

Comment:
here I think maybe we should check to see the number of bytes read matches what was passed-in?

Reply:
Don't think we'd have what to compare it to - the number of bytes returned from `io.Copy()` would be the number of uncompressed bytes, whereas this function receives only the compressed bytes.

Reply:
Actually, it turns out that Gzip stores the length and checksum of the uncompressed data, and the [gzip.Reader](https://golang.org/pkg/compress/gzip/#Reader) verifies that the number of uncompressed bytes matches that. So in case of any mismatch, an `ErrChecksum` will be propagated from the `gzip.Reader` up to the `io.Copy`.